crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,530 @@
+        Self { docs, ..self }
+    }
+
+    /// A representation of the type path of the set.
+    ///
+    /// Provides dynamic access to all methods on [`TypePath`].
+    pub fn type_path_table(&self) -> &TypePathTable {
+        &self.type_path
+    }
+
+    /// The [stable, full type path] of the set.
+    ///
+    /// Use [`type_path_table`] if you need access to the other methods on [`TypePath`].
+    ///
+    /// [stable, full type path]: TypePath
+    /// [`type_path_table`]: Self::type_path_table
+    pub fn type_path(&self) -> &'static str {
+        self.type_path_table().path()
+    }
+
+    /// The [`TypeId`] of the set.
+    pub fn type_id(&self) -> TypeId {
+        self.type_id
+    }
+
+    /// Check if the given type matches the set type.
+    pub fn is<T: Any>(&self) -> bool {
+        TypeId::of::<T>() == self.type_id
+    }
+
+    /// A representation of the type path of the value type.
+    ///
+    /// Provides dynamic access to all methods on [`TypePath`].
+    pub fn value_type_path_table(&self) -> &TypePathTable {
+        &self.value_type_path
+    }
+
+    /// The [`TypeId`] of the value.
+    pub fn value_type_id(&self) -> TypeId {
+        self.value_type_id
+    }
+
+    /// Check if the given type matches the value type.
+    pub fn value_is<T: Any>(&self) -> bool {
+        TypeId::of::<T>() == self.value_type_id
+    }
+
+    /// The docstring of this set, if any.
+    #[cfg(feature = "documentation")]
+    pub fn docs(&self) -> Option<&'static str> {
+        self.docs
+    }
+}
+
+/// An ordered set of reflected values.
+#[derive(Default)]
+pub struct DynamicSet {
+    represented_type: Option<&'static TypeInfo>,
+    hash_table: HashTable<Box<dyn Reflect>>,
+}
+
+impl DynamicSet {
+    /// Sets the [type] to be represented by this `DynamicSet`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given [type] is not a [`TypeInfo::Set`].
+    ///
+    /// [type]: TypeInfo
+    pub fn set_represented_type(&mut self, represented_type: Option<&'static TypeInfo>) {
+        if let Some(represented_type) = represented_type {
+            assert!(
+                matches!(represented_type, TypeInfo::Set(_)),
+                "expected TypeInfo::Set but received: {:?}",
+                represented_type
+            );
+        }
+
+        self.represented_type = represented_type;
+    }
+
+    /// Inserts a typed value into the set.
+    pub fn insert<V: Reflect>(&mut self, value: V) {
+        self.insert_boxed(Box::new(value));
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_hash(value: &Box<dyn Reflect>) -> u64 {
+        value.reflect_hash().expect(hash_error!(value))
+    }
+
+    #[allow(clippy::borrowed_box)]
+    fn internal_eq(value: &Box<dyn Reflect>) -> impl FnMut(&Box<dyn Reflect>) -> bool + '_ {
+        |other| {
+            value
+                .reflect_partial_eq(&**other)
+                .expect("Not the same type")
```

Comment:
`reflect_partial_eq` returns `false` if a type doesn't reflect `PartialEq`, not if it's different from `other`. In fact if they're different this returns `Some(false)` because they're clearly not equal.